### PR TITLE
rgw: add support for Swift's per storage policy statistics

### PR DESCRIPTION
**Pull request by rzarzynski**
*DNM* because the PR undergoes testing at the moment.

Fixes: http://tracker.ceph.com/issues/17932.

### CODE REVIEW

**Comment by rzarzynski** [comment]
Removing the *DNM* tag as Tempest nor s3-tests haven't found any regression here. Tempest additionally confirmed the new feature is operational.

CC: @yehudasa, @oritwas, @cbodley.

**Comment by yehudasa** [comment]
@rzarzynski I started looking at it, however, it seems slightly incompatible with wip-bucket-cleanup-3 that I want to merge in first. Can you look at rebasing this one on top of it?

**Comment by rzarzynski** [comment]
@yehudasa: sure, I work on that.

**Comment by oritwas** [comment]
@rzarzynski , please rebase

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
The branch has been tested in following Teuthology runs:
* http://pulpito.ceph.com/rzarzynski-2017-06-12_17:02:04-rgw-wip-rzarzynski-testing---basic-mira/,
* http://pulpito.ceph.com/rzarzynski-2017-06-13_13:39:20-rgw-wip-rzarzynski-testing---basic-mira/.

The results look clean. The two tests are red because of the known `rgw multisite test failures`. In the first log we can find:
```
2017-06-13T20:13:03.212 INFO:tasks.ceph.c2.osd.2.mira066.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.2-2424-g248209d/rpm/el7/BUILD/ceph-12.0.2-2424-g248209d/src/common/shared_cache.hpp: 105: FAILED assert(weak_refs.empty())
2017-06-13T20:13:03.262 INFO:tasks.ceph.c2.osd.2.mira066.stderr: ceph version  12.0.2-2424-g248209d (248209d231db64a4a9d3a4d6e002e13eeafe26eb) luminous (dev)
2017-06-13T20:13:03.262 INFO:tasks.ceph.c2.osd.2.mira066.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x110) [0xafa870]
2017-06-13T20:13:03.262 INFO:tasks.ceph.c2.osd.2.mira066.stderr: 2: (()+0x5030da) [0x60b0da]
2017-06-13T20:13:03.262 INFO:tasks.ceph.c2.osd.2.mira066.stderr: 3: (OSDService::~OSDService()+0x17c) [0x594bac]
2017-06-13T20:13:03.263 INFO:tasks.ceph.c2.osd.2.mira066.stderr: 4: (OSD::~OSD()+0x133) [0x5e31a3]
2017-06-13T20:13:03.263 INFO:tasks.ceph.c2.osd.2.mira066.stderr: 5: (OSD::~OSD()+0x9) [0x5e37e9]
2017-06-13T20:13:03.263 INFO:tasks.ceph.c2.osd.2.mira066.stderr: 6: (main()+0x2f48) [0x4dc1b8]
2017-06-13T20:13:03.263 INFO:tasks.ceph.c2.osd.2.mira066.stderr: 7: (__libc_start_main()+0xf5) [0xd5c0b35]
2017-06-13T20:13:03.263 INFO:tasks.ceph.c2.osd.2.mira066.stderr: 8: (()+0x46d1d6) [0x5751d6]
2017-06-13T20:13:03.263 INFO:tasks.ceph.c2.osd.2.mira066.stderr: NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
2017-06-13T20:13:03.264 INFO:tasks.ceph.c2.osd.2.mira066.stderr:2017-06-13 20:13:02.107604 323fd700 -1 osd.2 71 pgid 6.c has ref count of 2
2017-06-13T20:13:03.265 INFO:tasks.ceph.c2.osd.2.mira066.stderr:2017-06-13 20:13:02.107960 323fd700 -1 osd.2 71 pgid 6.24 has ref count of 2
2017-06-13T20:13:03.265 INFO:tasks.ceph.c2.osd.2.mira066.stderr:2017-06-13 20:13:02.108314 323fd700 -1 osd.2 71 pgid 6.1c has ref count of 2
2017-06-13T20:13:03.265 INFO:tasks.ceph.c2.osd.2.mira066.stderr:2017-06-13 20:13:02.108677 323fd700 -1 osd.2 71 pgid 6.12 has ref count of 2
2017-06-13T20:13:03.266 INFO:tasks.ceph.c2.osd.2.mira066.stderr:2017-06-13 20:13:02.109068 323fd700 -1 osd.2 71 pgid 6.26 has ref count of 2
2017-06-13T20:13:03.312 INFO:tasks.ceph.c2.osd.2.mira066.stderr:2017-06-13 20:13:03.206774 9630080 -1 leaked refs:
2017-06-13T20:13:03.313 INFO:tasks.ceph.c2.osd.2.mira066.stderr:dump_weak_refs 0xf6c6230 weak_refs: 65 = 0x36835340 with 18 refs
2017-06-13T20:13:03.313 INFO:tasks.ceph.c2.osd.2.mira066.stderr:dump_weak_refs 0xf6c6230 weak_refs: 66 = 0x33da20c0 with 43 refs
2017-06-13T20:13:03.313 INFO:tasks.ceph.c2.osd.2.mira066.stderr:dump_weak_refs 0xf6c6230 weak_refs: 69 = 0x34e88df0 with 7 refs
2017-06-13T20:13:03.313 INFO:tasks.ceph.c2.osd.2.mira066.stderr:dump_weak_refs 0xf6c6230 weak_refs: 70 = 0x36902610 with 3 refs
2017-06-13T20:13:03.313 INFO:tasks.ceph.c2.osd.2.mira066.stderr:dump_weak_refs 0xf6c6230 weak_refs: 71 = 0x33acc1f0 with 141 refs
```
The second one:
```
2017-06-13T17:27:48.388 INFO:tasks.ceph.c2.osd.0.mira035.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.2-2424-g248209d/rpm/el7/BUILD/ceph-12.0.2-2424-g248209d/src/common/shared_cache.hpp: 105: FAILED assert(weak_refs.empty())
2017-06-13T17:27:48.482 INFO:tasks.ceph.c2.osd.0.mira035.stderr: ceph version  12.0.2-2424-g248209d (248209d231db64a4a9d3a4d6e002e13eeafe26eb) luminous (dev)
2017-06-13T17:27:48.483 INFO:tasks.ceph.c2.osd.0.mira035.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x110) [0xafa870]
2017-06-13T17:27:48.483 INFO:tasks.ceph.c2.osd.0.mira035.stderr: 2: (()+0x5030da) [0x60b0da]
2017-06-13T17:27:48.483 INFO:tasks.ceph.c2.osd.0.mira035.stderr: 3: (OSDService::~OSDService()+0x17c) [0x594bac]
2017-06-13T17:27:48.483 INFO:tasks.ceph.c2.osd.0.mira035.stderr: 4: (OSD::~OSD()+0x133) [0x5e31a3]
2017-06-13T17:27:48.483 INFO:tasks.ceph.c2.osd.0.mira035.stderr: 5: (OSD::~OSD()+0x9) [0x5e37e9]
2017-06-13T17:27:48.483 INFO:tasks.ceph.c2.osd.0.mira035.stderr: 6: (main()+0x2f48) [0x4dc1b8]
2017-06-13T17:27:48.483 INFO:tasks.ceph.c2.osd.0.mira035.stderr: 7: (__libc_start_main()+0xf5) [0xd5c0b35]
2017-06-13T17:27:48.484 INFO:tasks.ceph.c2.osd.0.mira035.stderr: 8: (()+0x46d1d6) [0x5751d6]
2017-06-13T17:27:48.484 INFO:tasks.ceph.c2.osd.0.mira035.stderr: NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
2017-06-13T17:27:48.504 INFO:tasks.ceph.c2.osd.0.mira035.stderr:2017-06-13 17:27:48.385144 9630080 -1 leaked refs:
2017-06-13T17:27:48.504 INFO:tasks.ceph.c2.osd.0.mira035.stderr:dump_weak_refs 0xf6c2e20 weak_refs: 101 = 0x375be5d0 with 6 refs
```
@yehudasa: is there a possibility to have those changes in Luminous?

File: src/cls/user/cls_user.cc [hunk 1]
```diff
@@ -163,13 +163,18 @@ static int cls_user_set_buckets_info(cls_method_context_t hctx, bufferlist *in,
     CLS_LOG(20, "storing entry for key=%s size=%lld count=%lld",
             key.c_str(), (long long)update_entry.size, (long long)update_entry.count);
 
+    // Update bucket's placement rule info only when linking bucket, not on
+    // usage stats change. */
+    if (op.add) {
```
Comment (amitkumar50) [inline src/cls/user/cls_user.cc]: @rzarzynski How good would it would be to place 'if(op.add)` and `if (!op.add)` in if-else statement?

**Comment by rzarzynski** [comment]
Rebased.

**Comment by rzarzynski** [comment]
This PR has been retested in: http://pulpito.ceph.com/rzarzynski-2017-08-30_12:11:05-rgw-wip-rgw-swift_api_testing---basic-smithi/.

**Comment by rzarzynski** [comment]
jenkins retest this please (no log)

**Comment by mdw-at-linuxbox** [comment]
This change is required to have ceph pass "tempest" swift tests, and there are people would like to see that.  So yes, it would be good for this to make it to luminous.

**Comment by mdw-at-linuxbox** [comment]
Can you rebase this on something newer so that it passes "make check"?

**Comment by mattbenjamin** [comment]
@rzarzynski yes, please rebase, we'll target merge this week

**Comment by rzarzynski** [comment]
@mdw-at-linuxbox, @mattbenjamin: just rebased.

File: src/rgw/rgw_common.h [hunk 2]
```diff
@@ -1918,7 +1919,7 @@ struct RGWBucketEnt {
   }
 
   void encode(bufferlist& bl) const {
-    ENCODE_START(6, 5, bl);
+    ENCODE_START(7, 5, bl);
```
Comment (mattbenjamin) [inline src/rgw/rgw_common.h]: ack

File: src/cls/user/cls_user_types.h [hunk 3]
```diff
@@ -105,10 +105,15 @@ struct cls_user_bucket_entry {
...
+  /* The placement_rule is necessary to calculate per-storage-policy statics
+   * of the Swift API. Although the info available in RGWBucketInfo, we need
+   * to duplicate it here to not affect the performance of buckets listing. */
+  std::string placement_rule;
+
   cls_user_bucket_entry() : size(0), size_rounded(0), count(0), user_stats_sync(false) {}
 
   void encode(bufferlist& bl) const {
-    ENCODE_START(7, 5, bl);
+    ENCODE_START(8, 5, bl);
```
Comment (mattbenjamin) [inline src/cls/user/cls_user_types.h]: ack

File: src/cls/user/cls_user_types.h [hunk 4]
```diff
@@ -146,6 +152,8 @@ struct cls_user_bucket_entry {
       ::decode(user_stats_sync, bl);
     if (struct_v >= 7)
       ::decode(creation_time, bl);
+    if (struct_v >= 8)
```
Comment (mattbenjamin) [inline src/cls/user/cls_user_types.h]: ack

File: src/rgw/rgw_op.h [hunk 5]
```diff
@@ -582,34 +582,38 @@ class RGWBulkUploadOp::AlignedStreamGetter
 }; /* RGWBulkUploadOp::AlignedStreamGetter */
 
 
+struct RGWUsageStats {
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.h]: like this

File: src/rgw/rgw_rest.h [hunk 6]
```diff
@@ -678,6 +679,24 @@ static inline void dump_header_prefixed(struct req_state* s,
 }
 
 template <class... Args>
+static inline void dump_header_infixed(struct req_state* s,
+                                       const boost::string_ref& prefix,
+                                       const boost::string_ref& infix,
+                                       const boost::string_ref& sufix,
+                                       Args&&... args) {
+  char full_name_buf[prefix.size() + infix.size() + sufix.size() + 1];
```
Comment (mattbenjamin) [inline src/rgw/rgw_rest.h]: would love to someday remove sprintfs

**Pull request merged**